modules/docs/markdown/04-codegen/01-customisation/10-service-product.md

```diff
@@ -0,0 +1,217 @@
+We call this version a "service product" because it can be seen as the product
+of all the operations of the service.
+
+To generate a service product, annotate the service definition with
+
+`@generateServiceProduct`:
+
+```kotlin
+@SERVICE_PRODUCT_SPEC@
+```
+
+This will generate the following interface:
+
+```scala
+trait ExampleServiceProductGen[F[_, _, _, _, _]] {
+  def exampleOperation: F[ExampleInput, ExampleServiceOperation.ExampleOperationError, ExampleOutput, Nothing, Nothing]
+}
+```
+
+and the following implementation of `ServiceProduct`:
+
+```scala
+object ExampleServiceProductGen extends ServiceProduct[ExampleServiceProductGen]
+```
+
+You will be able to access the service product version of the service like this:
+
+```scala mdoc
+import smithy4s.example.product._
+
+ExampleServiceGen.serviceProduct
+```
+
+Or, more generically:
+
+```scala mdoc
+import smithy4s.ServiceProduct
+
+def productOf[Alg[_[_, _, _, _, _]]](mirror: ServiceProduct.Mirror[Alg]) = mirror.serviceProduct
+
+// example
+def exampleProduct = productOf(ExampleService)
+```
+
+With service products, you can call service methods without providing their inputs directly.
+
+Here are a couple ways you can use this as a library author:
+
+### Static description of services
+
+```scala mdoc
+import smithy4s.kinds.PolyFunction5
+
+type Describe[_, _, _, _, _] = String
+
+def descriptor[Alg[_[_, _, _, _, _]]](mirror: ServiceProduct.Mirror[Alg]): mirror.Prod[Describe] =
+  mirror
+    .serviceProduct
+    .mapK5(
+      mirror.serviceProduct.endpointsProduct,
+      new PolyFunction5[mirror.serviceProduct.service.Endpoint, Describe] {
+
+        override def apply[I, E, O, SI, SO](
+          fa: mirror.serviceProduct.service.Endpoint[I, E, O, SI, SO]
+        ): Describe[I, E, O, SI, SO] =
+          s"def ${fa.name}(input: ${fa.input.shapeId.name}): ${fa.output.shapeId.name}"
+
+      },
+    )
+
+// Usage
+
+val desc: String = descriptor(ExampleService).exampleOperation
+```
+
+### Non-linear input of operation
+
+```scala mdoc
+import smithy4s.ShapeId
+
+type Id[A] = A
+
+val impl: ExampleService[Id] =
+  new ExampleService[Id] {
+
+    override def exampleOperation(input: String): ExampleOutput = ExampleOutput(
+      s"Output for $input!"
+    )
+
+  }
+
+type ListClient[I, _, O, _, _] = List[I] => List[O]
+
+def listClient[Alg[_[_, _, _, _, _]], Prod[_[_, _, _, _, _]]](
+  impl: smithy4s.kinds.FunctorAlgebra[Alg, Id]
+)(
+  implicit sp: ServiceProduct.Aux[Prod, Alg]
+): Prod[ListClient] = sp
+  .mapK5(
+    sp.endpointsProduct,
+    new PolyFunction5[sp.service.Endpoint, ListClient] {
+
+      private val interp = sp.service.toPolyFunction(impl)
+
+      override def apply[I, E, O, SI, SO](
+        fa: sp.service.Endpoint[I, E, O, SI, SO]
+      ): List[I] => List[O] = _.map(in => interp(fa.wrap(in)))
+
+    },
+  )
+
+// Usage
+
+listClient(impl)( /* implicit scope problem here - TODO */ ExampleService.serviceProduct)
```

Comment:
**issue:** we're lacking something that would make this visible implicitly.

I tried the following (all combinations):

- making `serviceProduct` an implicit val in the interface
- making `serviceProduct` an implicit val in the generated code

I'm not sure why it's not being detected.

Reply:
> making serviceProduct an implicit val in the generated code 

where was the implicit located when you tried ?

Reply:
exactly where the non-implicit val is (companion object of `ExampleServiceGen`)